### ISSUE TITLE
sw_execute not finding CMAKE_CXX_COMPILER_ID

### ISSUE THREAD

**Issue by GitGoliath** [DanBloomberg/leptonica#702]
Hi,

I am new to Leptonica, I build on Windows 11 (22H2), with Visual Studio (Version 17.6.4), following the instruction. Install SW client, doing the setup, etc.

When running from VS GUI, the CMakeLists.txt, I get the following error in SWConfig.cmake line 166:

>> message(FATAL_ERROR "Compiler is not implemented: '${CMAKE_C_COMPILER_ID}' or '${CMAKE_CXX_COMPILER_ID}'")

I saw in the on git change log that 3 weeks ago the CXX where removed. Could be related...

I had to insert in the CMakeLists.txt the following: 

set(CMAKE_CXX_COMPILER_ID "MSVC")

as sw_execute seems to need it, and the current CMake script I have was not setting it. Possibly from the change 3 weeks ago. Note that the SW Client used is from
[sw-master-windows_x86_64-client.zip] (https://software-network.org/client/sw-master-windows_x86_64-client.zip) from yesterday
13-Jun-2023 21:33

So the current version does not seems to build from windows directly.

**Comment by GitGoliath** [comment]
An additonal note, adding set(CMAKE_C_COMPILER_ID "MSVC") did not resolve the issue. So SW seems to need CMAKE_CXX_COMPILER_ID.

**Comment by zdenop** [comment]
Not sure what you are doing but sw build seems to have no problem on the latest windows:

https://github.com/DanBloomberg/leptonica/actions/runs/5292787926/jobs/9580098275

**Comment by GitGoliath** [comment]
Humm, another question then, does the automatic test build use CMake, or is it using the Visual Studio CMake?

My setup is pretty clean, but I will try on a virtual machine to see if I can see the same issue. Will get back with the result.

Thanks

**Comment by zdenop** [comment]
@GitGoliath : maybe you can give you step-by-step description how did you get error?

@egorpugin: can you give your insight how to use sw?

**Comment by GitGoliath** [comment]
Hi,

Was able to reproduce from clean Windows 11 VMWare install.

Here is the steps by steps reproduction of the issue on a clean VMWare Virtual Machine (Windows 11 version 22H2, build 22621.1928:

1- From Clean Windows 11 latest update virtual machine, download Visual Studio Community 
2- Launch installation package with admin priviledge (VisualStudioSetup)
3- Select the following options:
    - ASP.NET and web development
    - Azure Development
    - Python development
    - Node.js development
    - .NET Multi-platform App UI development
    - .NET desktop development
    - Desktop development with C++
    - Mobile development with C++
    - Game development with Unity
    - Game Development with C++
    - Data storage and processing
    - Data science and analytical application
4- Download SW client, from https://software-network.org/client/    , version named sw-master-windows_x86_64-client
5- Unzip and copy SW.exe to "C:\SW Client"
6- Launch Command Prompt with Admin Privilege, go to "C:\SW Client", execute command "sw setup"
7- Go to "Advanced system settings" dialog (search it in taskbar), press "Environment Variables" button, in "User variables for User" click "new" and add "C:\SW Client"
8- Launch Visual Studio, and clone Leptonica repository from "https://github.com/DanBloomberg/leptonica.git" into "C:\Git\Leptonica" folder
9- CMake should automatically launch after cloning and I get the line 166 error in SWConfig.cmake from the sw_execute of the Leptonica CmakeLists.txt (line 233). 

Here is the output window:

1> CMake generation started for default configuration: 'x64-Debug'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Git\Leptonica\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"  -DCMAKE_MAKE_PROGRAM="C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja\ninja.exe" "C:\Git\Leptonica" 2>&1"
1> Working directory: C:\Git\Leptonica\out\build\x64-Debug
1> [CMake] -- Found SW: C:/SW Client/sw.exe  
1> [CMake] -- sw: processing dependencies
1> [CMake] CMake Error at C:/Users/VMLoc/.sw/storage/etc/sw/static/SWConfig.cmake:166 (message):
1> [CMake]   Compiler is not implemented: 'MSVC' or ''
1> [CMake] Call Stack (most recent call first):
1> [CMake]   CMakeLists.txt:213 (sw_execute)
1> [CMake] -- Configuring incomplete, errors occurred!
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Git\Leptonica\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"  -DCMAKE_MAKE_PROGRAM="C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja\ninja.exe" "C:\Git\Leptonica" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Git\Leptonica\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"  -DCMAKE_MAKE_PROGRAM="C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja\ninja.exe" "C:\Git\Leptonica" 2>&1"' returned with exit code: 1'.


From error window:


Error		CMake Error at C:/Users/VMLoc/.sw/storage/etc/sw/static/SWConfig.cmake:166 (message):
  Compiler is not implemented: 'MSVC' or ''		C:/Users/VMLoc/.sw/storage/etc/sw/static/SWConfig.cmake	166

**Comment by egorpugin** [comment]
Hi,

Please try to build from command line.
Post your command and output.

**Comment by egorpugin** [comment]
Yes, you should resolve this first.
Later you could check with IDE.

**Comment by GitGoliath** [comment]
Ok, I used CMake instead of the Microsoft "CMake" included with Visual Studio. Taken from https://cmake.org/download/.
I used the GUI version and configure and generated Leptonica in the build folder. The two CMake are not equal... :(

I was then able to open the Visual Studio (.sln) project generated. When building with latest master (2024/12/28), it seems that gif_lib.h is not found for the project "gifio_reg". There is even in the code a patch to redirect CMake to use GIF_INCLUDE_DIR to find it (seems to be missing in Windows based on comment of the patch). However GIF_INCLUDE_DIR is not set anywhere. To finaly have it compile I added an entry in CMake GUI interface for the path GIF_INCLUDE_DIR set to the location SW put the file (C:/Users/[user]/.sw/storage/pkg/20/75/c94f/src/sdir).

Now it compiles on my system. :)
